src/librbd/io/ImageRequestWQ.cc

```diff
@@ -311,6 +313,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
     trace.event("init");
   }
 
+  //block overlapping IOs
+  uint64_t tid = 0;
+  if (native_async && (m_image_ctx.non_blocking_aio || writes_blocked())) {
+    {
+      std::lock_guard locker{ordering_lock};
+      tid = ++m_last_tid;
+      ldout(cct, 20) << "queueing tid: " << tid << dendl;
```

Comment:
Nit: why can't this logic get moved down into `start_in_flight_io`?

Reply:
Added a new function `handle_overlapping_io` to handle this, since `start_in_flight_io` in it's current form is needed while unblocking IOs .